tls/s2n_fingerprint_ja4.c

```diff
@@ -0,0 +1,567 @@
+{
+    RESULT_ENSURE_REF(ch);
+
+    s2n_parsed_extension *extension = NULL;
+    RESULT_GUARD_POSIX(s2n_client_hello_get_parsed_extension(
+            S2N_EXTENSION_SIGNATURE_ALGORITHMS, &ch->extensions, &extension));
+    RESULT_ENSURE_REF(extension);
+
+    struct s2n_stuffer sig_algs = { 0 };
+    RESULT_GUARD_POSIX(s2n_stuffer_init_written(&sig_algs, &extension->extension));
+
+    uint8_t entry_bytes[S2N_JA4_IANA_ENTRY_SIZE] = { 0 };
+    struct s2n_stuffer entry = { 0 };
+    RESULT_GUARD_POSIX(s2n_blob_init(&entry.blob, entry_bytes, sizeof(entry_bytes)));
+
+    bool is_first = true;
+    RESULT_GUARD_POSIX(s2n_stuffer_skip_read(&sig_algs, sizeof(uint16_t)));
+    while (s2n_stuffer_data_available(&sig_algs)) {
+        uint16_t iana = 0;
+        RESULT_GUARD_POSIX(s2n_stuffer_read_uint16(&sig_algs, &iana));
+        if (s2n_is_grease_value(iana)) {
+            continue;
+        }
+        if (is_first) {
+            RESULT_GUARD(s2n_fingerprint_hash_add_char(hash, S2N_JA4_PART_DIV));
+        } else {
+            RESULT_GUARD_POSIX(s2n_stuffer_write_char(&entry, S2N_JA4_LIST_DIV));
+        }
+        RESULT_GUARD(s2n_stuffer_write_uint16_hex(&entry, iana));
+        RESULT_GUARD(s2n_fingerprint_hash_add_bytes(hash, entry_bytes,
+                s2n_stuffer_data_available(&entry)));
+        RESULT_GUARD_POSIX(s2n_stuffer_rewrite(&entry));
+        is_first = false;
+    }
+    return S2N_RESULT_OK;
+}
+
+/**
+ *= https://raw.githubusercontent.com/FoxIO-LLC/ja4/v0.18.2/technical_details/JA4.md#extension-hash
+ *# A 12 character truncated sha256 hash of the list of extensions, sorted by
+ *# hex value, followed by the list of signature algorithms, in the order that
+ *# they appear (not sorted).
+ */
+static S2N_RESULT s2n_fingerprint_ja4_c(struct s2n_fingerprint *fingerprint,
+        struct s2n_fingerprint_hash *hash, struct s2n_stuffer *output,
+        struct s2n_blob *extensions_count)
+{
+    RESULT_ENSURE_REF(fingerprint);
+
+    uint16_t extensions_count_value = 0;
+    RESULT_GUARD(s2n_fingerprint_ja4_extensions(hash, fingerprint->client_hello,
+            &fingerprint->workspace, &extensions_count_value));
+
+    /**
+     *= https://raw.githubusercontent.com/FoxIO-LLC/ja4/v0.18.2/technical_details/JA4.md#extension-hash
+     *# The signature algorithm hex values are then added to the end of the list
+     *# in the order that they appear (not sorted) with an underscore delimiting
+     *# the two lists.
+     *
+     *= https://raw.githubusercontent.com/FoxIO-LLC/ja4/v0.18.2/technical_details/JA4.md#extension-hash
+     *# If there are no signature algorithms in the hello packet,
+     *# then the string ends without an underscore and is hashed.
+     *
+     * s2n_fingerprint_ja4_sig_algs handles writing the underscore because we
+     * need to skip writing it if there are no signature algorithms.
+     */
+    s2n_result_ignore(s2n_fingerprint_ja4_sig_algs(hash, fingerprint->client_hello));
```

Comment:
Is there a reason why this result is ignored? It seems odd to potentially hash part of a malformed signature algorithms list without erroring. And this seems unique to signature algorithms.

Would it make sense to have `s2n_fingerprint_ja4_sig_algs` handle the no signature algorithms case, so this can error when the signature algorithms list is malformed?

Reply:
JA4 isn't clear on error handling. You're right that it's probably not worth trying to swallow errors other than a missing extension (and empty extension, for testing convenience). If the extension exists and contains malformed data, it'll probably be fine to fail.

Reply:
If you think it'd be better to avoid erroring, it might also work to bail out of `s2n_fingerprint_ja4_sig_algs()` if the list is malformed. The weird part to me was swallowing the error in addition to hashing part of the malformed list.